### ISSUE TITLE
Unable to parse AAAA record with IPv4 IP

### ISSUE THREAD

**Issue by lanrat** [miekg/dns#1107]
Some DNS servers respond with AAAA records that contain IPv4 addresses. This is not too unusual, and `dns.UnpackRR` works correctly on these records.

However, if this dns library receives one of these packets, unpacks it into a `RR`, and converts it to its string representation, it is unable to parse it back again.

The [way to encode these addresses](https://blog.ip2location.com/knowledge-base/ipv4-mapped-ipv6-address/) is to prefix them with `::ffff:` so `192.168.1.1` would become `::ffff:192.168.1.1`
RFC: https://tools.ietf.org/html/rfc4291#section-2.2

I would suggest treating AAAA records that contain IPv4 addresses as IPv4 addresses in IPv6 notation like the above example, instead of erroring.

Below is a example to demo this behavior with a real record I've come across:
```golang
package main

import (
	"encoding/hex"
	"log"

	"github.com/miekg/dns"
)

var rrHex = "036e733104696c706d026d7900001c000100015180001000000000000000000000ffff6e042d04"

/* rrHex is equivalent to:
rr := &dns.AAAA{
	Hdr: dns.RR_Header{
		Name:   "ns1.ilpm.my.",
		Rrtype: dns.TypeAAAA,
		Class:  dns.ClassINET,
		Ttl:    86400,
		Rdlength: 16,
	},
	AAAA: net.ParseIP("110.4.45.4"),
} */

func main() {
	log.Printf("Hex: %s", rrHex)
	rrBytes, err := hex.DecodeString(rrHex)
	if err != nil {
		log.Fatal(err)
	}
	rr, _, err := dns.UnpackRR(rrBytes, 0)
	if err != nil {
		log.Fatal(err)
	}
	rrString := rr.String()
	log.Printf("Parsed RR: %s", rrString)

	rr2, err := dns.NewRR(rrString)
	if err != nil { // err: dns: bad AAAA AAAA: "110.4.45.4" at line: 1:37
		log.Fatal(err)
	}
	log.Printf("Parsed RR from string: %s", rr2.String())
}

```

**Comment by lanrat** [comment]
If passing a IPv4 address in IPv6 notation to `net.ParseIP` only an IPv4 address is returned. example: https://play.golang.org/p/qt706imaofj

It looks like this case is already handled when converting AAAA records to strings, here (https://github.com/miekg/dns/blob/master/types.go#L1408) so it should also support paring records like this as well, or the RR String method should prefix `::FFFF:` when displaying.

**Comment by lanrat** [comment]
`"ns1.ilpm.my.     86400   IN      AAAA    ::ffff:110.4.45.4"` is successfully parsed by `dns.NewRR` as `"ns1.ilpm.my.     86400   IN      AAAA    110.4.45.4"`, but this can not be parsed back again.

It looks like there are two issues here:

1. The string representation for AAAA records should prefix the IP with `::ffff:` if it is an IPv4 IP
2. `dns.NewRR` should allow IPv4 strings to be used for `AAAA` records as the underlying go representation does not differentiate.

**Comment by gibson042** [comment]
Disagree on point 2, `ns1.ilpm.my. 86400 IN AAAA 110.4.45.4` is not valid and should be neither accepted nor produced.

**Comment by lanrat** [comment]
> Disagree on point 2, `ns1.ilpm.my. 86400 IN AAAA 110.4.45.4` is not valid and should be neither accepted nor produced.

I get your point here. 

I process a lot of zone files and just noticed this behavior. So I currently have hundreds of zone files generated with this library with the incorrect IPv6 formatting. 

I can fork this library, add the ability to read these malformed AAAA records, and re-parse and generate all of them, but I was hoping I would not need to do that.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Unable to parse AAA..." ]
>Disagree on point 2, ns1.ilpm.my. 86400 IN AAAA 110.4.45.4 is not valid and
>should be neither accepted nor produced.

excactly. This is a weird corner case and if you need IPv4 addresses you can use `A` for
that.

**Comment by lanrat** [comment]
`"ns1.ilpm.my.     86400   IN      AAAA    ::ffff:110.4.45.4"` is a valid AAAA record and is parsed by this library without error, but the string representation returned is `"ns1.ilpm.my.     86400   IN      AAAA    110.4.45.4"` which is incorrect.

#1108 offers a solution to this.

**Comment by lanrat** [comment]
> `"ns1.ilpm.my. 86400 IN AAAA ::ffff:110.4.45.4"` is a valid AAAA record and is parsed by this library without error, but the string representation returned is `"ns1.ilpm.my. 86400 IN AAAA 110.4.45.4"` which is incorrect.
> 
> #1108 offers a solution to this.

@miekg do you believe that the behavior described in this comment is a bug or desired behavior?

These types of responses are possible in DNS64 as described in [RFC 6052](https://tools.ietf.org/html/rfc6052).

My understanding is that the current behavior is a derivation from [RFC 4291](https://tools.ietf.org/html/rfc4291#section-2.2).

**Comment by gibson042** [comment]
The textual representation of the data portion of the AAAA resource record is defined by [RFC 3596](https://tools.ietf.org/html/rfc3596#section-2.4) as "the textual representation of an IPv6 address as defined in [RFC 3513]". RFC 3513 was obsoleted by RFC 4291, but the relevant "Text Representation of Addresses" section had no relevant changes.

`::ffff:110.4.45.4` is a valid textual represenation of AAAA RDATA; `110.4.45.4` is not.

**Comment by lanrat** [comment]
> The textual representation of the data portion of the AAAA resource record is defined by [RFC 3596](https://tools.ietf.org/html/rfc3596#section-2.4) as "the textual representation of an IPv6 address as defined in [RFC 3513]". RFC 3513 was obsoleted by RFC 4291, but the relevant "Text Representation of Addresses" section had no relevant changes.
> 
> `::ffff:110.4.45.4` is a valid textual represenation of AAAA RDATA; `110.4.45.4` is not.

Exactly. Currently if you create a AAAA record with this library for `::ffff:110.4.45.4` and then call `String()` on it, the IPv6 address is printed as `110.4.45.4`.

**Comment by tmthrgd** [comment]
Considering prefixing `::ffff:` is already done for `APLPrefix`, I would be in favour for doing that. @lanrat Do you want to open a PR and then we can see how invasive such a change would be?

**Comment by lanrat** [comment]
@tmthrgd I already submitted #1108 that uses the same method used in `APLPrefix` to prefix the IP with `::ffff:`.
However it was closed with no comment which confused me.

**Comment by tmthrgd** [comment]
@lanrat I hadn't seen that. I suppose things can get overlooked pretty easily. I'll leave it closed for now and see what @miekg thinks.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Unable to parse AAA..." ]
>@lanrat I hadn't seen that. I suppose things can get overlooked pretty easily.
>I'll leave it closed for now and see what @miekg thinks.

if the std lib supports it we should, if not then we should bent over backwards to make
this happen (and remove anything works around that we have).

**Comment by lanrat** [comment]
> [ Quoting <notifications@github.com> in "Re: [miekg/dns] Unable to parse AAA..." ]
> @lanrat I hadn't seen that. I suppose things can get overlooked pretty easily. I'll leave it closed for now and see what @miekg thinks.
> if the std lib supports it we should, if not then we should bent over backwards to make this happen (and remove anything works around that we have).

If this is the case, then it should allowed to parse IPv4 addressees in `AAAA` records. Otherwise, this library can output records that it itself can not re-parse. Which is what lead me to this issue in the first place.

**Comment by lanrat** [comment]
I would prefer using something like my pr #1108 to print `AAAA` as IPv6 always over allowing IPv4 addresses in `AAAA` records. This keep this library more inline with not allowing "garbage" in, and not outputting any "garbage" of its own.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/dns] Unable to parse AAA..." ]
>    [ Quoting notifications@github.com in "Re: [miekg/dns] Unable to parse
>    AAA..." ]
>    @lanrat I hadn't seen that. I suppose things can get overlooked pretty
>    easily. I'll leave it closed for now and see what @miekg thinks.
>    if the std lib supports it we should, if not then we should bent over
>    backwards to make this happen (and remove anything works around that we
>    have).
>
>If this is the case, then it should allowed to parse IPv4 addressees in AAAA
>records. Otherwise, this library can output records that it itself can not
>re-parse. Which is what lead me to this issue in the first place.

that seems fair

**Comment by lanrat** [comment]
> that seems fair

@miekg to confirm, are you saying you would rather print the AAAA records in question as an IPv4 address over displaying them as an IPv6 address?

I'm happy to submit a PR for this, but it seems a little out of line with this library not outputting any "garbage" just due to an artifact of golang's underlying IP implementation.

**Comment by lanrat** [comment]
Another reason not to output AAAA records with IPv4 addresses in IPv4 notation:

When using bind to parse a zonefile created with this library, bind will throw the following error with this library's current implementation: `Bind: :185: near '175.182.228.35': bad IPv6 address`.

However, if you prepend the IPv4 IP with `:ffff::` then the zonefile loads.

I'm not saying we should define what is the correct behavior by what bind wants, but I think that this helps illustrate that the current behavior does result in this library outputting invalid records. And that the fix should be on the display `.String()` side and not in the parser.

**Comment by lanrat** [comment]
I don't believe this issue is solved. Its still possible for this library to parse an IPv6 address, emit an IPv4 address, and then be unable to re-parse it.

**Comment by lanrat** [comment]
For anyone still blocked on this. Looks like it was finally fixed in d00c6728a4024bed201b2ede34f2ecd47e05b1e0.
